src/provider/provider_coarse.c

```diff
@@ -477,16 +756,28 @@ static enum umf_result_t coarse_memory_provider_alloc(void *provider,
                        coarse_provider->alloc_size);
             }
 
+            rv = allocated_blocks_add(coarse_provider->allocated_blocks, curr);
+            if (rv) {
+                return UMF_RESULT_ERROR_OUT_OF_HOST_MEMORY;
+            }
+
             // Try to merge new empty block with the next one.
-            block_merge_with_next(new_block, &coarse_provider->block_list);
+            new_block =
+                block_merge_with_next(coarse_provider->free_blocks, new_block,
+                                      &coarse_provider->block_list);
+
+            rv = free_blocks_add(coarse_provider->free_blocks, new_block);
+            if (rv) {
+                return UMF_RESULT_ERROR_OUT_OF_HOST_MEMORY;
+            }
 
             if (util_mutex_unlock(coarse_provider->lock) != 0) {
                 return UMF_RESULT_ERROR_MEMORY_PROVIDER_SPECIFIC;
             }
 
             assert(debug_check(coarse_provider));
             return UMF_RESULT_SUCCESS;
-        } else if (!curr->used && (curr->size == size)) {
+        } else if ((curr->size == size)) {
```

Comment:
you could assert that curr->used == false

Reply:
Done